{
 "cells": [
  {
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "\n",
    "# Aufgabe 2\n",
    "\n",
    "Im Rahmen der Materialverwaltung soll ein Programm zur automatischen Analyse von Materialeingängen und Materialabgängen implementiert werden. So soll aus einer großen Liste von Eingängen und Abgängen erkannt werden wie der neue Materialbestand ist. Außerdem soll die durchschnittliche Höhe der Materialbewegungen berechnet werden, sowie der größte Ein- oder Abgang bestimmt werden.\n",
    "\n",
    "Implementieren Sie nun die Funktion `analyse_bestand`. Diese Funktion erhält als Parameter eine Liste `eingehend_ausgehend`, welche in Form von Integer-Werten die Eingänge und Abgänge auf das Konto angibt (in etwa `[+500, -200, +70, -65]`).  \n",
    "Als Ergebnis soll der Materialbestand(`bestand`) als Integer-Wert zurückgegeben werden.  \n",
    "Die durchschnittliche Höhe(`durchschnitt`) der Eingänge und Abgänge eines Materials.  \n",
    "Die höchste Transaktion (`hoechste_transaktion`). Hierbei sollen sowohl Abgänge als auch Eingänge berücksichtigt werden.  \n",
    "</div>\n",
    "<div class=\"alert alert-block alert-info\">\n",
    "\n",
    "### Hilfe\n",
    "Sie können Ihre Lösung mit den automatisierten [Tests](#Automatische-Tests) weiter unten im Dokument überprüfen.<br>\n",
    "Eine Einführung in die `For-Schleifen` in Python finden Sie [hier](../../50_loops/for_loop.ipynb).\n",
    "</div>"
   ],
   "cell_type": "markdown",
   "metadata": {
    "nbgrader": {
     "grade": false,
     "grade_id": "cell-5af2db4a976d2c43",
     "locked": true,
     "schema_version": 1,
     "solution": false
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "nbgrader": {
     "grade": false,
     "grade_id": "cell-3785c4bc8f420267",
     "locked": false,
     "schema_version": 1,
     "solution": true
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "    \n",
    "    bestand = 0\n",
    "    hoechste_transaktion = 0\n",
    "    for transaktion in eingehend_ausgehend:\n",
    "        bestand += transaktion\n",
    "        # Prüfe ob die Transaktion die höchste ist. Wenn ja, speichere den Wert als \"hoechste_transaktion\"\n",
    "    \n",
    "    # Berechne den Rückgabewert \"durchschnitt\"\n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "nbgrader": {
     "grade": false,
     "grade_id": "cell-c847824f224f66ce",
     "locked": true,
     "schema_version": 1,
     "solution": false
    }
   },
   "source": [
    "  ***\n",
    "  # Automatische Tests\n",
    "  Ab hier finden Sie einige automatisierte Tests um Ihre Lösungen zu überprüfen. Um Ihre Lösung zu prüfen führen Sie bitte zuerst die Zelle mit Ihrer Lösung und danach die Zelle mit dem zugehörigen Test aus.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append(\"../..\")\n",
    "from nose.tools import assert_equal, assert_not_equal\n",
    "from utils import * \n",
    "\n",
    "try:\n",
    "    incoming_outgoing = [+519, -250, -19, -12, +234]\n",
    "    Erg = analyse_bestand(incoming_outgoing)\n",
    "    expected_out = 472\n",
    "    assert_equal(Erg[0], expected_out)\n",
    "    print(\"[OK] Es wurde der richtige Bestand berechnet!\")\n",
    "except AssertionError as e:\n",
    "    print(\"[FEHLER] Es wurde\",Erg ,\"berechnet , obwohl \", expected_out, \"richtig wäre.\")\n",
    "    print(\"       \",e)\n",
    "except Exception as e:\n",
    "    print(\"[FEHLER]\", friendly_exceptions.msg(e))\n",
    "\n",
    "try:\n",
    "    incoming_outgoing = [+99999999, -1234567, -9876543]\n",
    "    Erg = analyse_bestand(incoming_outgoing)\n",
    "    expected_out = 88888889\n",
    "    assert_equal(Erg[0], expected_out)\n",
    "    print(\"[OK] Es wurde der richtige Bestand berechnet!\")\n",
    "except AssertionError as e:\n",
    "    print(\"[FEHLER] Es wurde\",Erg ,\"berechnet , obwohl \", expected_out, \"richtig wäre.\")\n",
    "    print(\"       \",e)\n",
    "except Exception as e:\n",
    "    print(\"[FEHLER]\", friendly_exceptions.msg(e))\n",
    "\n",
    "try:\n",
    "    incoming_outgoing = [-12, +20]\n",
    "    Erg = analyse_bestand(incoming_outgoing)\n",
    "    expected_out = 8\n",
    "    assert_equal(Erg[0], expected_out)\n",
    "    print(\"[OK] Es wurde der richtige Bestand berechnet!\")\n",
    "except AssertionError as e:\n",
    "    print(\"[FEHLER] Es wurde\",Erg ,\"berechnet , obwohl \", expected_out, \"richtig wäre.\")\n",
    "    print(\"       \",e)\n",
    "except Exception as e:\n",
    "    print(\"[FEHLER]\", friendly_exceptions.msg(e))\n",
    "\n",
    "try:\n",
    "    incoming_outgoing = [+300, -15, +230, -123, +987]\n",
    "    Erg = analyse_bestand(incoming_outgoing)\n",
    "    expected_out = 275.8\n",
    "    assert_equal(Erg[1], expected_out)\n",
    "    print(\"[OK] Es wurde der richtige durchschnittliche Bestand berechnet!\")\n",
    "except AssertionError as e:\n",
    "    print(\"[FEHLER] Es wurde der falsche Durschnitt von\",Erg ,\"berechnet , obwohl \", expected_out, \"richtig wäre.\")\n",
    "    print(\"       \",e)\n",
    "except Exception as e:\n",
    "    print(\"[FEHLER]\", friendly_exceptions.msg(e))\n",
    "\n",
    "try:\n",
    "    incoming_outgoing = [+200, -200, +200, -200]\n",
    "    Erg = analyse_bestand(incoming_outgoing)\n",
    "    expected_out = 0\n",
    "    assert_equal(Erg[1], expected_out)\n",
    "    print(\"[OK] Es wurde der richtige durchschnittliche Bestand berechnet!\")\n",
    "except AssertionError as e:\n",
    "    print(\"[FEHLER] Es wurde der falsche Durschnitt von\",Erg ,\"berechnet , obwohl \", expected_out, \"richtig wäre.\")\n",
    "    print(\"       \",e)\n",
    "except Exception as e:\n",
    "    print(\"[FEHLER]\", friendly_exceptions.msg(e))\n",
    "\n",
    "try:\n",
    "    incoming_outgoing = [+15]\n",
    "    Erg = analyse_bestand(incoming_outgoing)\n",
    "    expected_out = 15\n",
    "    assert_equal(Erg[1], expected_out)\n",
    "    print(\"[OK] Es wurde der richtige durchschnittliche Bestand berechnet!\")\n",
    "except AssertionError as e:\n",
    "    print(\"[FEHLER] Es wurde der falsche Durschnitt von\",Erg ,\"berechnet , obwohl \", expected_out, \"richtig wäre.\")\n",
    "    print(\"       \",e)\n",
    "except Exception as e:\n",
    "    print(\"[FEHLER]\", friendly_exceptions.msg(e))\n",
    "\n",
    "try:\n",
    "    incoming_outgoing = [+100, -26, -23, +210, +234]\n",
    "    Erg = analyse_bestand(incoming_outgoing)\n",
    "    expected_out = 234\n",
    "    assert_equal(Erg[2], expected_out)\n",
    "    print(\"[OK] Es wurde die richtige höchste Bewegung erkannt!\")\n",
    "except AssertionError as e:\n",
    "    print(\"[FEHLER] Es wurde die falsche Höchste Bewegung von\",Erg ,\"berechnet , obwohl \", expected_out, \"richtig wäre.\")\n",
    "    print(\"       \",e)\n",
    "except Exception as e:\n",
    "    print(\"[FEHLER]\", friendly_exceptions.msg(e))\n",
    "\n",
    "try:\n",
    "    incoming_outgoing = [+200, -200, +200, -200]\n",
    "    Erg = analyse_bestand(incoming_outgoing)\n",
    "    expected_out = 200\n",
    "    assert_equal(Erg[2], expected_out)\n",
    "    print(\"[OK] Es wurde die richtige höchste Bewegung erkannt!\")\n",
    "except AssertionError as e:\n",
    "    print(\"[FEHLER] Es wurde die falsche Höchste Bewegung von\",Erg ,\"berechnet , obwohl \", expected_out, \"richtig wäre.\")\n",
    "    print(\"       \",e)\n",
    "except Exception as e:\n",
    "    print(\"[FEHLER]\", friendly_exceptions.msg(e))\n",
    "\n",
    "try:\n",
    "    incoming_outgoing = [+15]\n",
    "    Erg = analyse_bestand(incoming_outgoing)\n",
    "    expected_out = 15\n",
    "    assert_equal(Erg[2], expected_out)\n",
    "    print(\"[OK] Es wurde die richtige höchste Bewegung erkannt!\")\n",
    "except AssertionError as e:\n",
    "    print(\"[FEHLER] Es wurde die falsche Höchste Bewegung von\",Erg ,\"berechnet , obwohl \", expected_out, \"richtig wäre.\")\n",
    "    print(\"       \",e)\n",
    "except Exception as e:\n",
    "    print(\"[FEHLER]\", friendly_exceptions.msg(e))\n",
    "\n",
    "try:\n",
    "    incoming_outgoing = [+300, -500, -23, +210, +234]\n",
    "    Erg = analyse_bestand(incoming_outgoing)\n",
    "    expected_out = -500\n",
    "    assert_equal(Erg[2], expected_out)\n",
    "    print(\"[OK] Es wurde die richtige höchste Bewegung erkannt!\")\n",
    "except AssertionError as e:\n",
    "    print(\"[FEHLER] Es wurde die falsche Höchste Bewegung von\",Erg ,\"berechnet , obwohl \", expected_out, \"richtig wäre.\")\n",
    "    print(\"       \",e)\n",
    "except Exception as e:\n",
    "    print(\"[FEHLER]\", friendly_exceptions.msg(e))"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Create Assignment",
  "file_extension": ".py",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.2"
  },
  "mimetype": "text/x-python",
  "name": "python",
  "npconvert_exporter": "python",
  "pygments_lexer": "ipython3",
  "version": 3
 },
 "nbformat": 4,
 "nbformat_minor": 2
}